### PR TITLE
Fix high-order output for vtu in parallel

### LINKED ISSUE #14275: `DataOut::write_vtu_in_parallel`+ `DataOutBase::VtkFlags::write_higher_order_cells` + `dim==3`: distorted cells in Paraview
While using `DataOut::write_vtu_in_parallel` (with #processes>1) together with `DataOutBase::VtkFlags::write_higher_order_cells=true`, `n_subdivisons>1` and `dim==3`, I obtain distorted cells in Paraview (v5.10.0), see right picture in the following figure. To me it looks like a numbering issue, but I am not sure. Does someone have a suspicion what might go wrong here or have similar problems? It would be really great to make use of this flag, since otherwise the file size of the output is almost doubled.

bug

I've used the following code to produce the results:

```cpp
#include <deal.II/distributed/tria.h>
#include <deal.II/dofs/dof_handler.h>

#include <deal.II/base/function_signed_distance.h>
#include <deal.II/dofs/dof_tools.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/mapping_q1.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/lac/generic_linear_algebra.h>
#include <deal.II/numerics/data_out.h>
#include <deal.II/numerics/vector_tools.h>

#include <iostream>

using namespace dealii;

using VectorType = LinearAlgebra::distributed::Vector<double>;

template <int dim>
void test(const unsigned int fe_degree,
          const bool write_higher_order_cells = true) {
  parallel::distributed::Triangulation<dim> triangulation(MPI_COMM_WORLD);

  GridGenerator::hyper_cube(triangulation, 0, 1);
  triangulation.refine_global(2);

  MappingQ1<dim> mapping;
  FE_Q<dim> fe(fe_degree);

  DoFHandler<dim> dof_handler(triangulation);
  dof_handler.distribute_dofs(fe);

  VectorType solution(dof_handler.n_dofs());

  IndexSet locally_relevant_dofs;
  DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);

  solution.reinit(dof_handler.locally_owned_dofs(), locally_relevant_dofs,
                  MPI_COMM_WORLD);

  dealii::VectorTools::interpolate(
      mapping, dof_handler, Functions::SignedDistance::Sphere<dim>(), solution);

    DataOutBase::VtkFlags flags;
    flags.write_higher_order_cells = write_higher_order_cells;

    DataOut<dim> data_out;
    data_out.set_flags(flags);

    data_out.add_data_vector(dof_handler, solution, "solution");

    data_out.build_patches(mapping, fe_degree);

    // write vtu in parallel
    std::string output("test_vtu_parallel_dim" + std::to_string(dim) + "_" +
                       std::to_string(fe_degree) + "_" +
                       std::to_string(write_higher_order_cells) + ".vtu");
    data_out.write_vtu_in_parallel(output, MPI_COMM_WORLD);
}

int main(int argc, char *argv[]) {
  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);

  test<3>(2);
  test<3>(2, false /*write_higher_order_cells*/);

  return 0;
}
```

FYI @peterrum @nmuch

### PR DESCRIPTION
**Pull request by peterrum**
closes #14275 

This was a funny bug. Indeed, there has been recently a change in ordering of lines in the vtk standard (https://github.com/Kitware/VTK/blob/7a0b92864c96680b1f42ee84920df556fc6ebaa3/Documentation/release/dev/node-numbering-change-for-VTK_LAGRANGE_HEXAHEDRON.md). But the old numbering is still supported (version number 2.1 and less). However, Paraview has problems with old versions of vtk if multiple "pieces" are in the same file (what we do in `write_vtu_in_parallel`).

I would suggest to increase the version of the vtu version to 2.2 (if high-order output is requested) and adopt the new ordering. We could also do the same for vtk  (versions 5.2); but there are some additional changes in the standard. 

FYI @mschreter

### CODE REVIEW

File: doc/news/changes/incompatibilities/20220918Munch [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Updated: The version of the vtu ouput has been increased to 2.2
```
Comment (kronbichler) [inline doc/news/changes/incompatibilities/20220918Munch:1]:
```suggestion
Updated: The version of the vtu output has been increased to 2.2
```

File: doc/news/changes/incompatibilities/20220918Munch [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+if high-order ouput is requested.
```
Comment (kronbichler) [inline doc/news/changes/incompatibilities/20220918Munch:2]:
```suggestion
if high-order output is requested.
```

File: doc/news/changes/incompatibilities/20220918Munch [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+This was necessary to fix visualization issues in Paraview.
```
Comment (kronbichler) [inline doc/news/changes/incompatibilities/20220918Munch:3]:
```suggestion
This is necessary to fix visualization issues in Paraview.
```

**Review by kronbichler — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by mschreter** [comment]
@peterrum great, I've verified that the artifacts are gone (also for `write_vtu_with_pvtu_record()`). Thanks a lot!

File: source/base/data_out_base.cc [hunk 4]
```diff
@@ -5653,7 +5671,7 @@ namespace DataOutBase
     // now for the cells
     out << "CELLS " << n_cells << ' ' << n_points_and_n_cells << '\n';
     if (flags.write_higher_order_cells)
-      write_high_order_cells(patches, vtk_out);
+      write_high_order_cells(patches, vtk_out, true);
```
Comment (tjhei) [inline source/base/data_out_base.cc:5674]:
```suggestion
      write_high_order_cells(patches, vtk_out, /* legacy_format = */ true);
```

File: source/base/data_out_base.cc [hunk 5]
```diff
@@ -964,12 +965,20 @@ namespace
...
+   * https://github.com/Kitware/VTK/blob/7a0b92864c96680b1f42ee84920df556fc6ebaa3/Documentation/release/dev/node-numbering-change-for-VTK_LAGRANGE_HEXAHEDRON.md
+   *
    */
   int
   vtk_point_index_from_ijk(const unsigned                 i,
                            const unsigned                 j,
                            const unsigned                 k,
-                           const std::array<unsigned, 3> &order)
+                           const std::array<unsigned, 3> &order,
+                           const bool                     legacy)
```
Comment (tjhei) [inline source/base/data_out_base.cc:981]:
```suggestion
                           const bool                     legacy_format)
```

File: source/base/data_out_base.cc [hunk 6]
```diff
@@ -6042,7 +6064,7 @@ namespace DataOutBase
     out << "    <DataArray type=\"Int32\" Name=\"connectivity\" format=\""
         << ascii_or_binary << "\">\n";
     if (flags.write_higher_order_cells)
-      write_high_order_cells(patches, vtu_out);
+      write_high_order_cells(patches, vtu_out, false);
```
Comment (tjhei) [inline source/base/data_out_base.cc:6067]:
```suggestion
      write_high_order_cells(patches, vtu_out, /* legacy_format = */ false);
```

**Review by tjhei — APPROVED** [review]

**Comment by peterrum** [comment]
> @peterrum great, I've verified that the artifacts are gone now (also for write_vtu_with_pvtu_record()). Thanks a lot for your work!

@mschreter `write_vtu_with_pvtu_record()` calls `write_vtu_in_parallel()`, which calls `write_vtu()`. Thanks for checking!

@tjhei I have incorporated your suggestions.

**Review by tjhei — APPROVED** [review]

**Pull request merged**